scipy/stats/_bws_test.py

```diff
@@ -0,0 +1,167 @@
+import numpy as np
+from scipy import stats
+from functools import partial
+from collections import namedtuple
+from ._axis_nan_policy import _axis_nan_policy_factory
+
+
+def bws_statistic(x, y, alternative, method):
+    Ri, Hj = np.sort(x), np.sort(y)
+    n, m = Ri.shape[-1], Hj.shape[-1]
+    i, j = np.arange(1, n+1), np.arange(1, m+1)
+
+    if alternative == 'two-sided':
+        if method == 'naive':
+            Bx_num = (Ri - (m + n) / n * i) ** 2
+            By_num = (Hj - (m + n) / m * j) ** 2
+            Bx_den = i / (n + 1) * (1 - i / (n + 1)) * m * (m + n) / n
+            By_den = j / (m + 1) * (1 - j / (m + 1)) * n * (m + n) / m
+        else:
+            Bx_num = (Ri - (m + n)/n * i) * abs(Ri - (m + n)/n * i)
+            By_num = (Hj - (m + n)/n * j) * abs(Hj - (m + n)/n * j)
+            Bx_den = i / (n + 1) * (1 - i / (n + 1)) * \
+                m * (m + n + 1) / (n + 2)
+            By_den = j / (m + 1) * (1 - j / (m + 1)) * \
+                n * (m + n + 1) / (m + 2)
+
+        Bx = 1 / n * np.sum(Bx_num / Bx_den)
+        By = 1 / m * np.sum(By_num / By_den)
+
+        B = (Bx + By) * 0.5
+    else:
+        if method == 'naive':
+            Bx_num = (Ri - (m + n + 1) / (n + 1) * i) ** 2
+            By_num = (Hj - (m + n) / m * j) ** 2
+        else:
+            Bx_num = (Ri - (m + n)/n * i) * abs(Ri - (m + n)/n * i)
+            By_num = (Hj - (m + n)/n * j) * abs(Hj - (m + n)/n * j)
+
+        Bx_den = i / (n + 1) * (1 - i / (n + 1)) * m * (m + n) / n
+        By_den = j / (m + 1) * (1 - j / (m + 1)) * n * (m + n) / m
+
+        Bx = 1 / n * np.sum(Bx_num / Bx_den)
+        By = 1 / m * np.sum(By_num / By_den)
+
+        if alternative == 'greater':
+            B = (Bx - By) * 0.5
+        else:
+            B = (By - Bx) * 0.5
+
+    return B
+
+
+BWSResult = namedtuple('BWSResult', ('statistic', 'pvalue'))
+
+
+@_axis_nan_policy_factory(BWSResult, n_samples=2)
```

Comment:
This can be done in a follow-up. Please remove it.

Reply:
> Do you mean by _axis_nan_policy_factory or namedtuple? Why not keep it for now?

`_axis_nan_policy_factory` is almost a magic bullet, but not quite. It needs to be configured correctly, and whether it is configured correctly needs to be tested. There will be enough to work on in this PR without that extra complexity, so I am suggesting removing it to make this PR as simple as possible.